### PR TITLE
fix Julia performance

### PR DESCRIPTION
**Pull request by oscardssmith**
Putting the code in a function means that you get type stability dropping the time down to 145 ms on my computer.

### CODE REVIEW

File: src/leibniz.jl [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
-for i in 2:(rounds + 2)
-    x *= -1
-    pi += x / (2 * i - 1)
+    return pi*4
 end
 
-pi *= 4
-print(pi)
+file = open("rounds.txt")
+rounds = parse(Int64, strip(readstring(file)))
```
Comment (giordano) [inline src/leibniz.jl:14]:
```suggestion
rounds = parse(Int64, readchomp(joinpath(@__DIR__, "rounds.txt")))
```

**Comment by niklas-heer** [comment]
Thank you for your contributition @oscardssmith. 🙏 
Also thank you @giordano. 👍

**Comment by oscardssmith** [comment]
Thanks for merging this. Any chance you can update the readme?

**Comment by niklas-heer** [comment]
@oscardssmith I'm currently working on a new version using [earthly](https://earthly.dev/) and a [new custom Go application](https://github.com/niklas-heer/speed-comparison/tree/next/scbench) to be able to isolate test runs better. Having a big Docker image doesn't work very well. The new version will be able to use the best Docker image for the language.

I'll keep you updated.

If you want to track progress: https://github.com/niklas-heer/speed-comparison/tree/next

**Comment by oscardssmith** [comment]
thanks!

**Comment by niklas-heer** [comment]
@oscardssmith thank you for your patience. I'm done reworking. The results are very different. But it also has been over 4 years since the last run, so it's probably also due to changes in the languages itself.
Here are the results:
plot

**Comment by giordano** [comment]
@niklas-heer how are you running the benchmark?  In particular, are you avoiding including compilation time by running the function twice?  On my 7-year old laptop this function takes about 4ms, when removing compilation latency:
```julia
julia> @time f(rounds) # first run, it includes compilation
  0.013784 seconds (31.09 k allocations: 1.708 MiB, 72.31% compilation time)
3.1415916535917745

julia> @time f(rounds) # second run, it doesn't include compilation
  0.004539 seconds (1 allocation: 16 bytes)
3.1415916535917745

julia> using BenchmarkTools # use package for more accurate benchmarking

julia> @btime f(rounds)
  3.793 ms (1 allocation: 16 bytes)
3.1415916535917745
```

**Comment by giordano** [comment]
#21 should make the first run less bad (actually pretty decent out-of-the-box for me).

**Comment by niklas-heer** [comment]
@giordano I'm running it in Docker containers on my MacBook. I wouldn't look too much at the absolute value of the results, but more on the relative performance. 

> Are you including whole Julia startup time by any chance?
Yes, the Go application measures the time it takes a given command to be executed. [Here](https://github.com/niklas-heer/speed-comparison/blob/3becd7711062a27dddef712eae70efb283b90bfe/Earthfile#L122) is an example:
```bash
./scbench "julia leibniz.jl" -i $iterations -l "julia --version" --export json --lang "Julia"
```
This means `julia leibniz.jl` is measured as a whole. There is not really a way around it if it should be automated.

What do you think?

**Comment by giordano** [comment]
The [`BenchmarkTools.jl`](https://github.com/JuliaCI/BenchmarkTools.jl) package (but you need to install it with `using Pkg; Pkg.add("BenchmarkTools")`) does the looping business under the hood.  For example
```console
$ julia --startup-file=no -L leibniz.jl -E 'print("\033[2K\033[1G"); using BenchmarkTools; @belapsed(f($rounds))'
0.003792468
```
The printing is to remove the value of pi printed by the script

**Comment by MasonProtter** [comment]
> 



> Yes, the Go application measures the time it takes a given command to be executed. [Here](https://github.com/niklas-heer/speed-comparison/blob/3becd7711062a27dddef712eae70efb283b90bfe/Earthfile#L122) is an example:
> 
> ```shell
> ./scbench "julia leibniz.jl" -i $iterations -l "julia --version" --export json --lang "Julia"
> ```
> 
> This means `julia leibniz.jl` is measured as a whole. There is not really a way around it if it should be automated.
> 
> What do you think?

I think this goes against what you said you do on the README, which states
> > Are the compile times included in the measurements?
> 
> No they are not included, because when running the program in the real world this would also be done before.

You are including Julia's internal compilation / startup time, as well as the JIT compilation of the actual function in question.

**Pull request merged**